### DISCUSSION TITLE
Without Remote - SSH, what's the point?

### DISCUSSION THREAD

**Discussion by wxnarwhal** [coder/code-server#3988] [Q&A] [answered]
Strong question, I know.  I'll introduce myself by saying I'm not a programmer, I just play one at work.  I'm a systems admin, I manage servers, networks and storage.  I've set up GitLab for our programmers in the past, but never used it as a user.  With all the things that changed in the last year I have no programmers left, and I'm spending about 10 hours a week doing some PHP programming myself.  I have a general concept of requirements.  Never work on the live server, so I've made staging servers.  Create branches, and don't work on unrelated features in the same branch.  VS Code on my desktop has helped greatly with me learning this workflow.  However, I still have one lingering issue, and that is that I work from home and 2 different offices.  I don't like duplicating my environment on 4 different computers, so code-server seems like a nice solution (I've since learned about Setting Sync, which may resolve the original reason I tried code-server).  However, after getting it installed, and installing Remote - SSH plugin, it doesn't work (and I've seen all the bugs about why, so I'm not asking why it doesn't work). 

A couple of options I've thought of:

- Install code-server on every staging server.  Doesn't solve my problem because that'll end up being way more servers than my desktops.
- Mount staging servers with SSHFS? Seems like I'd lose a huge amount of features of working on a remote server if I have to make code-server think I'm working locally.

These both seem like bad ideas, but if code-server is otherwise incapable of connecting to remote environments, how do people use it like the desktop version of VS Code? Maybe a better question is, do I totally misunderstand code-servers purpose?

**Comment by lg-nineoneninethreenine** [comment]
I read your description and you said that you work from home and at 2 different offices. Try getting the latest `code-server` on your computer/server, and MAKE SURE that it's connected to WiFi or Ethernet. Then, run this:
```bash
code-server --link
```
Note that this is in beta, as the Readme says. Open the GitHub Auth Link and login. That will allocate you a domain like so:
```
https://<server-hostname>.<github-username>.cdr.co
```
So, you run the server and if you open that, you will have to log in to GitHub and then, you get into your code-server. You can also set codenames, like @jsjoeio said, like so:
```bash
code-server --link <codename>
```
Replace `<codename>` with your code name of choice, like, for example: 
```bash
code-server --link server1
```
which will allocate a URL that is:
```
https://server1.im-coder-lg.cdr.co
```
**Note**: `Remember` that the `--link` flag is in beta stage, so things may happen. For example(real problem):

It allows latency below 100/75 ms but not above 100ms. I opened a discussion for a solution.

And, an answer for the title:

- You can use the beta flag, as I told above
- You can also customise the "hostname" to a codename
- Allows multiple servers(?)

Hope this solves the problem.

And yeah, you can use the built-in SSH prompt to control the other servers or activate code-server.

**Reply by wxnarwhal** [reply]
Thanks for replying, but this answer really confuses me.  How does this help code-server connect to a remote environment?  Unless I misunderstand the documentation, --link just creates a URL to access code-server where it's already installed.

**Reply by lg-nineoneninethreenine** [reply]
Ah, now you are explaining. Let me tell this: code-server is itself on a remote/local environment, why do you want to add more to it? Instead, if you connect to the servers via SSH and initialize `--link`, then all you need is a couple of tabs.

**Reply by wxnarwhal** [reply]
You might see it as adding, I see it as subtracting.  I want one install I can use from anywhere for anything.  Maybe code-server isn't the solution.  But code-server's self-described goal is "code on any device with a consistent development environment."  This is exactly what I wanted, to reduce my 4 environments to 1 "consistent development environment.".  Same plugins, same settings, same color schemes.  But if I'm working on 15 projects and I have to install locally on 15 staging servers, now I have 15 environments, far worse than my original problem of 4. There's no conceivable way 15 deployments could be considered "a consistent development environment."  And the problem gets worse if I add programmers.  We can't all share the same password.  Maybe I can create a new service to spin it up on a new port with a new password?  But if so, and I grow to 5 programmers for my 15 projects, now I'm up to 45 deployments.  I don't see how this is manageable or sustainable. Beyond one single user on one single server, I can't see how code-server achieves what it claims to achieve.  And that's fine.  If it's not for me, then it's not for me.  And if the developers' intentions are different, more power to you.

**Reply by lg-nineoneninethreenine** [reply]
Well, idk about that. /ping @jsjoeio

**Comment by jsjoeio** [comment]
Great questions! Let's see if I can address a few of them.

> "code on any device with a consistent development environment."

You can achieve this with code-server by installing on one machine and being able to access it from anywhere! 

To keep things simple, let's say I [deploy code-server to Railway](https://www.youtube.com/watch?v=IWfkkaY5fXs). As long as I do all my development in my code-server instance on Railway, I get "a consistent development environment".

To "code on any device", all I have to do is open a browser and navigate to my code-server Railway deployment and boom, I can write code on my phone, tablet, desktop or laptop. 

That is the value/purpose I see myself in code-server. 

> And the problem gets worse if I add programmers. We can't all share the same password. Maybe I can create a new service to spin it up on a new port with a new password? But if so, and I grow to 5 programmers for my 15 projects, now I'm up to 45 deployments.

And you're right, code-server does not officially support multiple users in one environment. There are folks who have made it happen and shared in the discussions. [Here is one](https://github.com/cdr/code-server/discussions/3606) for example.

If you are looking for a more complex setup and want to make it easier to spin up environments for a team of programmers, then I would point you to [our product](https://coder.com/).

I hope that helps! Happy to elaborate on any of these points as well. I'll also cc @bpmct who may be able to share more.

**Reply by lg-nineoneninethreenine** [reply]
Why not use the Coder Trial? I read the discussion and discovered that it has VSCode, JetBrains IDE, and Remote SSH included. Try that and tell here.

**Comment by bpmct** [comment, marked as answer]
This is a super interesting thread and brings up a few points. Thanks for the background info too @wxnarwhal!

If I'm understanding correctly, you'd like to have a single cloud development environment (code-server) which is able to connect and develop on 15+ staging servers for different projects or platforms?

For a bit of context, here are a few ways code-server is being used with multiple projects and even users, despite not supporting the workflow you mentioned. Note: when I mention `server` below, it could either be a VM or a container.

We often see a single code-server used to develop multiple (15+) projects at once, but then that single server needs to have all of the dependencies for the projects. In this case, the environment is less of a "staging" server that hosts a single service, and more accurately a replacement for a developer's laptop, but in the cloud. As more developers join the a project, they get 1 server that contains all the dev dependencies (container or VM image), or they are responsible for installing stuff themselves (like locally) if the consistency isn't an important important. In this case, a real staging server may still necessary, but it is deployed on pushes via CI, not developed on.

And yep, you're correct. If code-server supported remote extensions, then code-server a could be a very useful "portal" into developing inside existing staging servers versus just being used to develop on the local host. Mounting staging servers via SSHFS would work but you would probably loose some extension support and debugging.

A different thing we're seeing is a "1 server per user per project" model in order to achieve the consistency of a staging server, with the server being very close to production. Each of these servers have a unique code-server deployment and sometimes even a SSH server to access via other IDEs. In order to give developers a good experience on all these servers, settings sync can be used or [even a dotfiles repo](https://dotfiles.github.io/) to specify a favorite shell, vim settings, etc. This results in A LOT of servers. It also means you may have to write a script to run on various servers to get the developer experience the way you want. Leveraging containers, Kubernetes, or a cloud development platform (like what we're doing at [Coder](https://coder.com/docs) makes managing multiple images, users, dotfiles, and workspaces much more manageable. Coder relies on a lot of Kubernetes features to do this and has also some custom stuff, such as injecting code-server into a container and managing the user's dotfiles repo.

Note Coder also supports one-project-per-workspace OR multiple-projects-per-workspace, it's un-opinionated in that way.

Coder currently only provisions workspaces as pods in a Kubernetes cluster. They can be accessed with code-server, another web IDE in the image, or even via SSH (so VS Code Remote SSH can be used from a local machine). We're researching other provisioning methods (e.g creating VMs or custom hooks) so that different types of "workspaces" or staging servers can be provisioned and managed according to your needs or workflow..

For your specific workflow TODAY, I would probably suggest one of these:
1. installing multiple code-servers on each staging server you have and leverage dotfiles repos and settings sync to get the settings you need ([Coder](https://coder.com/docs) or a similar tool can be a centralized control plane to access these from) OR
1. developing with all your projects on a single server and leave the staging server for CI OR
1. add SSH keys to all your staging servers and mostly use VS Code Remote instead of code-server to develop on the remote servers. code-server can be launched occasionally you want to use an iPad or open it up to an unfamiliar device

Curious your thoughts on this, especially from an outside perspective. We're definitely in an interesting time for remote development and I think would be super nice to get [remote extensions support on code-server](https://github.com/cdr/code-server/issues/1315).

**Reply by wxnarwhal** [reply]
Thanks for all the info.  Coder is actually what led me to code-server.  I figured I could save money and learn more if I hosted it myself.  I see now that Coder has a lot of custom options that aren't naively a part of code-server.

**Reply by bpmct** [reply]
Yeah Coder is actually a self-hosted option as well. If you find yourself managing a lot of workspaces/code-servers at once, I'd recommend it. Coder workspaces also support connecting from remote SSH, you would just run a command on your local client(s) to connect. Send me a DM on [our Slack](https://cdr.co/join-community) or ben at coder dot com. And I'll send you a license :)
